crates/red_knot_python_semantic/resources/mdtest/suppressions/type-ignore.md

```diff
@@ -0,0 +1,83 @@
+# Suppressing errors with `type: ignore`
+
+Type check errors can be suppressed by a `type: ignore` comment on the same line as the violation.
+
+## Simple `type: ignore`
+
+```py
+a = 4 + test  # type: ignore
+```
+
+## In parenthesized expression
+
+```py
+a = (
+    4 + test  # type: ignore
+)  # fmt: skip
+```
+
+## Before opening parentheses
+
+A suppression that applies to all errors before the openign parentheses.
+
+```py
+a: Test = (  # type: ignore
+  5
+)  # fmt: skip
```

Comment:
I'm a bit confused how this test is working. I see you use the end of the diagnostic range for finding applicable suppressions. In this case the diagnostic spans the entire statement, so the end of the diagnostic range would be on the third line. But the suppression range would be just the first line. So why does this suppression apply? (To be clear, I think it _should_ apply, I just don't understand how it works. I could debug and figure it out, but faster to ask you :) )

Reply:
This is just a bad example. There's only one error: the unresolved reference to Test. The declared type is unknown, so the assignment itself is fine. 

I think a better example would be to change the right side to Test too and add and acknowledge that error with an error pragma comment

Reply:
Oh, of course!

But does that mean that with the current implementation, in a case where the diagnostic range does span multiple lines:

```py
x: str = (
    1 + 3
)  # type: ignore
```

You would be required to suppress this on the last line (as shown), not on the first (after the opening paren)?

I find that a little counter-intuitive; I would expect to be able to suppress a multi-line diagnostic on the line where it starts. (Imagine for example a diagnostic covering an entire `def` or `class` statement: I'd want to suppress that with a `type: ignore` on the `def` or `class` line, not on the final line of the function or class body, which would be pretty weird.) How does this work in pyright and mypy?

Reply:
I'd have to look but there's the other parenthesis case where mypy doesn't do anything fancy

For classes and other suites. I don't think we should ever use the entire statement range. We should instead report the diagnostic on the name

Reply:
It looks like mypy and pyright both allow you to put the `type: ignore` on _any_ line within the diagnostic range:

*  https://pyright-play.net/?strict=true&code=IYLgBAzgLgTmC8YAUAoM6wEYwGoxowCYUBKFFAD3GjkSXQGIwoBPABwFNwBLAcwDsA9jA4F02PMTIoW1WAmRisuRs3ZcwfISKVTyLQnNqKMynLtWtOPAcNHSAXkYWpTEsFMvqb2jkA
* https://mypy-play.net/?mypy=latest&python=3.12&gist=2352369ed2274c23a8e54605a1307f16

(The difference between them in this example is that pyright includes the opening and closing parentheses in the diagnostic range; mypy does not.)

I think we will probably want to match this behavior, otherwise we are just creating an unnecessary barrier for migration as some people's `type: ignore` comments that worked on mypy/pyright will stop working for us.

Reply:
> It looks like mypy and pyright both allow you to put the type: ignore on any line within the diagnostic range:

I've to take another look but i'm a bit concerned about using intersection here because it would mean that an ignore comment in the body of a statement would ignore any error in that statement's clause. E.g. 

```
class ThisClassHasAnError:
	def test():
		a + 3  # type: ignore
```

We don't want that the `type: ignore` ignores any errors from the enclosing function or class!

Reply:
Sooo, this is all slightly more subtle ;)

Your examples don't work with Red Knot because we report a different error range:

* Red Knot: Reports the entire assignment as invalid (the statement!)
* Mypy: Reports the range of the right side (the assigned value)
* Pyright: Reports the range of the right side

So red knot is the weird one here.

To add another data point. Ruff does the opposite of what I have here. It searches by the range's start position https://play.ruff.rs/280de058-f694-4f93-8321-46474a0aa541

Overall, I'm worried about allowing suppression comments anywhere inside an expression. Let's say you have

```py
a = (
	some_call(a, b, c)   # type: ignore 
		+ d
		+ e
		+ b
)
```

The `type: ignore` now disables not just an error related to the call but any error inside the binary expression. I find this very surprising. It gets worse the more nested your expressions are. 

I could see a middle ground where we allow any suppression comment on the same line as the diagnostic range.

Reply:
My understanding is that Mypy also uses the start and end line numbers. 

Mypy checks if the error is ignored on any `origin.lines`

https://github.com/python/mypy/blob/1f9317f593dc41a2805a3093e2e1890665485e76/mypy/errors.py#L493-L505

And origin is defined as

https://github.com/python/mypy/blob/1f9317f593dc41a2805a3093e2e1890665485e76/mypy/errors.py#L98-L101

So I think we should do the same for now

Reply:
> We don't want that the `type: ignore` ignores any errors from the enclosing function or class!

Well, we agreed that we shouldn't have diagnostics spanning an entire function or class anyway. But yes, the point about full intersection potentially leading to too much suppression makes sense. It is interesting that this still seems to be what pyright does (at least that's how I interpret https://pyright-play.net/?strict=true&code=IYLgBAzgLgTmC8YAUAoM6wEYwGoxowCYUBKFFAD3GjkSXQGIwoBPABwFNwBLAcwDsA9jA4F02HGLDEyKFtVgJkUiY2bsuYPkJFSZ5FoQW1lGLLj1rWnHgOGjZAL2NLUZiXtJWNtnRyA ). Though I'm having trouble constructing any case where it's ambiguous -- pyright seems to mostly avoid "nested" diagnostics by using narrow enough ranges, and having erroring operations return Unknown which then causes the outer operation to no longer be an error.